src/main/java/org/openhab/tools/analysis/checkstyle/PackageExportsNameCheck.java

```diff
@@ -0,0 +1,147 @@
+ * @author Svilen Valkanov - Refactored the check
+ */
+public class PackageExportsNameCheck extends AbstractStaticCheck {
+    private static final String CORRECT_NAMING_OF_NOT_EXPORTED_PACKAGES_MESSAGE = "The package %s"
+            + " should be marked as \"internal\" if it will not be exported.";
+
+    private static final String PACKAGE_PATTERN = "[a-zA-Z0-9\\._-]*";
+
+    private final Logger logger = LoggerFactory.getLogger(PackageExportsNameCheck.class);
+
+    private String[] sourceDirectories;
+    private String[] excludedPackages;
+
+    /**
+     * Sets the configuration property for source directories.
+     */
+    public void setSourceDirectories(String[] sourceDirectories) {
+        this.sourceDirectories = sourceDirectories;
+    }
+
+    /**
+     * Sets the configuration property for excluding packages.
+     */
+    public void setExcludedPackages(String[] excludePackages) {
+        this.excludedPackages = excludePackages;
+    }
+
+    public PackageExportsNameCheck() {
+        setFileExtensions(MANIFEST_EXTENSION);
+    }
+
+    @Override
+    protected void processFiltered(File manifestFile, List<String> lines) throws CheckstyleException {
+        BundleInfo bundleInfo = parseManifestFromFile(manifestFile);
+        Set<String> uniqueManifestExports = bundleInfo.getExports().stream().map(export -> export.getName())
+                .collect(Collectors.toSet());
+
+        File projectDirectory = manifestFile.getParentFile().getParentFile();
+        Path projectDirectoryPath = projectDirectory.toPath();
+
+        Set<String> sourcePackages = new HashSet<>();
+        try {
+            for (String sourcePath : sourceDirectories) {
+                Path relativeSourcePath = Paths.get(sourcePath);
+                Path sourceDirectoryPath = projectDirectoryPath.resolve(relativeSourcePath);
+                sourcePackages.addAll(getFilteredPackagesFromSourceDirectory(sourceDirectoryPath));
+            }
+
+            sourcePackages.removeAll(uniqueManifestExports);
+            for (String packageName : sourcePackages) {
+                log(0, String.format(CORRECT_NAMING_OF_NOT_EXPORTED_PACKAGES_MESSAGE, packageName));
+            }
+        } catch (IOException e) {
+            logger.error("Problem occurred while processing directories. "
+                    + "The check will exit without logging any warnings!", e);
+        }
+    }
+
+    /**
+     * Filter and return the packages from the source directory. Only not excluded packages will be returned.
+     *
+     * @param sourcePath - The full path of the source directory
+     * @return {@link Set } of {@link String }s with the package names.
+     * @throws IOException if an I/O error is thrown while visiting files
+     */
+    private Set<String> getFilteredPackagesFromSourceDirectory(Path sourcePath) throws IOException {
+        final Set<String> packages = new HashSet<String>();

```

Comment:
Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`